File: platforms/ios-consumer/SparrowInvest/Models/DividendSummary.swift
//
//  DividendSummary.swift
//  SparrowInvest
//
//  Dividend tracking models
//

import Foundation

// MARK: - Dividend Record

struct DividendRecord: Codable, Identifiable {
    let id: String
    let fundCode: String
    let fundName: String
    let amount: Double
    let unitsHeld: Double
    let dividendPerUnit: Double
    let recordDate: Date
    let paymentDate: Date
    let status: DividendStatus

    init(
        id: String = UUID().uuidString,
        fundCode: String,
        fundName: String,
        amount: Double,
        unitsHeld: Double,
        dividendPerUnit: Double,
        recordDate: Date,
        paymentDate: Date,
        status: DividendStatus = .paid
    ) {
        self.id = id
        self.fundCode = fundCode
        self.fundName = fundName
        self.amount = amount
        self.unitsHeld = unitsHeld
        self.dividendPerUnit = dividendPerUnit
        self.recordDate = recordDate
        self.paymentDate = paymentDate
        self.status = status
    }
}

// MARK: - Dividend Status

enum DividendStatus: String, Codable {
    case announced = "Announced"
    case pending = "Pending"
    case paid = "Paid"
}

// MARK: - Dividend Summary

struct DividendSummary: Codable {
    var financialYear: String
    var totalReceived: Double
    var projectedAnnual: Double
    var dividendYield: Double
    var records: [DividendRecord]
    var nextExpectedDate: Date?

    static var empty: DividendSummary {
        DividendSummary(
            financialYear: "FY 2025-26",
            totalReceived: 0,
            projectedAnnual: 0,
            dividendYield: 0,
            records: [],
            nextExpectedDate: nil
        )
    }

    var recentRecords: [DividendRecord] {
        records.sorted { $0.paymentDate > $1.paymentDate }.prefix(5).map { $0 }
    }

    var monthlyAverage: Double {
        guard !records.isEmpty else { return 0 }
        return totalReceived / 12
    }
}

File: platforms/android-consumer/app/src/main/java/com/sparrowinvest/app/data/model/DividendSummary.kt
package com.sparrowinvest.app.data.model

import kotlinx.serialization.Serializable

enum class DividendStatus(val displayName: String) {
    ANNOUNCED("Announced"),
    PENDING("Pending"),
    PAID("Paid")
}

@Serializable
data class DividendRecord(
    val id: String,
    val fundCode: String,
    val fundName: String,
    val amount: Double,
    val unitsHeld: Double,
    val dividendPerUnit: Double,
    val recordDate: String,
    val paymentDate: String,
    val status: String = "Paid" // Announced, Pending, Paid
)

@Serializable
data class DividendSummary(
    val financialYear: String = "FY 2025-26",
    val totalReceived: Double = 0.0,
    val projectedAnnual: Double = 0.0,
    val dividendYield: Double = 0.0,
    val records: List<DividendRecord> = emptyList(),
    val nextExpectedDate: String? = null
) {
    companion object {
        val empty = DividendSummary()
    }

    val recentRecords: List<DividendRecord>
        get() = records.sortedByDescending { it.paymentDate }.take(3)
}
